### PR TITLE
stake-address registration-certificate: do not allow key-reg-deposit-amt before conway

### LINKED ISSUE #504: stake-address registration-certificate has the flag --key-reg-deposit-amt, it should not.
legacy command 

`cardano-cli stake-address registration-certificate 
`
and era based 

`cardano-cli babbage stake-address registration-certificate `

have the flag `--key-reg-deposit-amt` but it is really not needed on Babbage. This is purely cosmetic since the resulting certificate does conform to the Babbage cddl. In other words, using this flag does not have any impact but might be confusing, since the deposit in Babbage era should be supplied on the pre-conway fashion: when balancing the transaction.   So it would be good to remove the flag for Babbage era and legacy

---

Clarification [here on Slack](https://input-output-rnd.slack.com/archives/C031GFLC3DF/p1702302426416769): `--key-reg-deposit-amt` should only be present in `conway`.

### PR DESCRIPTION
**Pull request by smelc**
# Changelog

```yaml
- description: |
    `stake-address registration-certificate`: remove flag `--key-reg-deposit-amt` from all eras except conway
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - improvement    # QoL changes e.g. refactoring
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

Fixes https://github.com/input-output-hk/cardano-cli/issues/504

# Checklist

- [X] Commit sequence broadly makes sense and commits have useful messages
- [X] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [X] Self-reviewed the diff

### CODE REVIEW

**Review by carlhammann — COMMENTED** [review]
If I understand #504, and the function [`createRegistrationCertRequirements`](https://github.com/input-output-hk/cardano-cli/blob/694782210c6d73a1b5151400214ef691f6f3ecb0/cardano-cli/src/Cardano/CLI/EraBased/Run/StakeAddress.hs#L146) correctly, we can (should!) go even further. The `--key-reg-deposit-amt` flag should be present _only_ on the Conway command line (and the legacy one, where we can't avoid it...), shouldn't it?

File: cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs [hunk 1]
```diff
@@ -86,28 +86,26 @@ pStakeAddressRegistrationCertificateCmd :: ()
   -> Maybe (Parser (StakeAddressCmds era))
 pStakeAddressRegistrationCertificateCmd era = do
   forEraInEonMaybe era $ \sbe ->
-    caseShelleyToBabbageOrConwayEraOnwards
-      (\shelleyToBabbage -> subParser "registration-certificate"
-            $ Opt.info
-                ( StakeAddressRegistrationCertificateCmd (shelleyToBabbageEraToShelleyBasedEra shelleyToBabbage)
+    caseShelleyToAlonzoOrBabbageEraOnwards
+      (\shelleyToAlonzo ->
+            make ( StakeAddressRegistrationCertificateCmd (shelleyToAlonzoEraToShelleyBasedEra shelleyToAlonzo)
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs]: You have `sbe` in the scope already, so you may want to reuse it.
```suggestion
            make ( StakeAddressRegistrationCertificateCmd sbe
```

File: cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs [hunk 2]
```diff
@@ -86,28 +86,26 @@ pStakeAddressRegistrationCertificateCmd :: ()
...
+              babbageBuilder si = StakeAddressRegistrationCertificateCmd babbageOrConway si Nothing
+              conwayBuilder si lovelace = StakeAddressRegistrationCertificateCmd babbageOrConway si (Just lovelace)
+          in make $ case babbageOnwards of
+               BabbageEraOnwardsBabbage -> babbageBuilder <$> pStakeIdentifier <*> pOutputFile
+               BabbageEraOnwardsConway ->  conwayBuilder <$> pStakeIdentifier <*> pKeyRegistDeposit <*> pOutputFile
       )
       sbe
    where
      desc = Opt.progDesc "Create a stake address registration certificate"
+     make parser = subParser "registration-certificate"  (Opt.info parser desc)
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs]: This code is a bit hard to follow and reason about. How about something simpler, using `forEraInEonMaybe`:

```suggestion
      (\_babbageOnwards -> do
          let mpKeyRegistDeposit =
                sequenceA $ forEraInEonMaybe
                  @ConwayEraOnwards
                  era
                  (const pKeyRegistDeposit)

          make $ StakeAddressRegistrationCertificateCmd sbe
            <$> pStakeIdentifier
            <*> mpKeyRegistDeposit
            <*> pOutputFile
      )
      sbe
   where
     desc = Opt.progDesc "Create a stake address registration certificate"
     make parser = subParser "registration-certificate"  (Opt.info parser desc)
```

File: cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs [hunk 3]
```diff
@@ -86,28 +86,26 @@ pStakeAddressRegistrationCertificateCmd :: ()
...
-      (\conwayOnwards -> subParser "registration-certificate"
-            $ Opt.info
-                ( StakeAddressRegistrationCertificateCmd (conwayEraOnwardsToShelleyBasedEra conwayOnwards)
-                    <$> pStakeIdentifier
-                    <*> fmap Just pKeyRegistDeposit
-                    <*> pOutputFile
-                )
-            desc
+      (\babbageOnwards ->
+          let babbageOrConway = babbageEraOnwardsToShelleyBasedEra babbageOnwards
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs]: This is unnecessary, there's `sbe` in scope already.
```suggestion

```

**Review by carbolymer — COMMENTED** [review]
Couple of code style comments.

**Comment by CarlosLopezDeLara** [comment]
> legacy

That's correct. It's a conway only flag.

File: cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs (hunk 1 above)
Comment (smelc) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs]: Applied in new version :+1:

File: cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs (hunk 2 above)
Comment (smelc) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs]: I went for something simpler in the new version, can you have another look?

File: cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs (hunk 3 above)
Comment (smelc) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs]: Applied in new version :+1:

File: cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs [hunk 4]
```diff
@@ -87,27 +87,18 @@ pStakeAddressRegistrationCertificateCmd :: ()
 pStakeAddressRegistrationCertificateCmd era = do
   forEraInEonMaybe era $ \sbe ->
     caseShelleyToBabbageOrConwayEraOnwards
-      (\shelleyToBabbage -> subParser "registration-certificate"
-            $ Opt.info
-                ( StakeAddressRegistrationCertificateCmd (shelleyToBabbageEraToShelleyBasedEra shelleyToBabbage)
-                    <$> pStakeIdentifier
-                    <*> optional pKeyRegistDeposit
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs:94]: Why not `pure Nothing` for Shelley to Babbage??

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]

File: cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs (hunk 4 above)
Comment (smelc) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/StakeAddress.hs:94]: Hum yeah really smart, I hadn't thought of lifting over the value type rather than the `Parser` type :+1: 

Done the change.

**Review by Jimbo4350 — APPROVED** [review]
LGTM! Just squash the commits.

**Comment by smelc** [comment]
image
